### PR TITLE
Fetch vbios version from sysfs

### PR DESCRIPTION
**Pull request by superm1**
For dGPUs this adds the VBIOS version into the device created by the optionrom plugin.
For APUs this adds the VBIOS version into a device that hangs off the APU.

For example on APU:

```
└─Ryzen 5 PRO 6650U with Radeon Graphics:
  │   Device ID:          4bde70ba4e39b28f9eab1628f9dd6e6244c03027
  │   Vendor:             AMD
  │   GUIDs:              b9a2dd81-159e-5537-a7db-e7101d164d3f ← cpu
  │                       22f9ecf4-588d-5c0a-8326-6ebff3655c6d ← CPUID\PRO_0&FAM_19
  │                       52f8f9af-1ca9-5352-bef4-ceb232c888a5 ← CPUID\PRO_0&FAM_19&MOD_44
  │                       e94372a3-3ffb-5d1c-a579-c415b7313e52 ← CPUID\PRO_0&FAM_19&MOD_44&STP_1
  │   Device Flags:       • Internal device
  │
  ├─Graphics Processing Unit (GPU):
  │     Device ID:        2287ee908298807e95f89a9b7c1a435c1a634e33
  │     Current version:  113-REMBRANDT-X37
  │     Vendor:           Advanced Micro Devices, Inc. [AMD/ATI] (PCI:0x1002)
  │     GUIDs:            817e27b3-d1bb-591a-a1cc-e039d0a60be6 ← PCI\VEN_1002&DEV_1681
  │                       a57d78cd-3738-5c25-9bae-60208129eae4 ← PCI\VEN_1002&DEV_1681&REV_D2
  │                       9b7d6019-9008-55ab-8512-0e2851542b8a ← PCI\VEN_1002&DEV_1681&SUBSYS_17AA22F1
  │                       5c1de11d-7e81-5e75-b8f8-0c5102c64756 ← PCI\VEN_1002&DEV_1681&SUBSYS_17AA22F1&REV_D2
  │     Device Flags:     • Internal device
  │
  └─System Management Unit (SMU):
        Device ID:        db0330716216c629bb2c07256e5d018f499eb6ce
        Summary:          Microcontroller used within CPU/APU program 4
        Current version:  69.55.0
        Vendor:           AMD
        GUID:             79307ae6-a2ea-52e1-bf56-6abbaf3547ad ← /sys/devices/platform/AMDI0007:00
        Device Flags:     • Internal device
```

```
Type of pull request:

- [x] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupdplugin/fu-udev-device.c [hunk 1]
```diff
@@ -442,6 +463,17 @@ fu_udev_device_probe(FuDevice *device, GError **error)
 		}
 	}
 
+	/* if the device is a dGPU try to fetch it from vbios_version */
+	if (g_strcmp0(priv->subsystem, "pci") == 0 && priv->class == 0x30000 &&
```
Comment (hughsie) [inline libfwupdplugin/fu-udev-device.c:467]: are the device classes enumerated? It seems like something we should have as a #define somewhere.

File: plugins/amdgpu/fu-amdgpu-device.c [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * Copyright (C) 2023 Advanced Micro Devices Inc.
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-amdgpu-device.h"
```
Comment (hughsie) [inline plugins/amdgpu/fu-amdgpu-device.c:11]: i think the plugin should be amd-gpu and have the source file prefix of fu-amd-gpu TBH.

File: plugins/amdgpu/fu-amdgpu-device.c [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	fu_device_add_parent_guid(device, "cpu");
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_INTERNAL);
+	return fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "pci", error);
+}
+
+static void
+fu_amdgpu_device_init(FuAmdGpuDevice *self)
+{
+	fu_device_set_name(FU_DEVICE(self), "Graphics Processing Unit (GPU)");
+	fu_device_add_icon(FU_DEVICE(self), "computer");
```
Comment (hughsie) [inline plugins/amdgpu/fu-amdgpu-device.c:40]: really `computer`?

Comment (superm1) [reply plugins/amdgpu/fu-amdgpu-device.c:40]: I was between that and monitor/display. Opinions?

File: plugins/amdgpu/fu-amdgpu-device.c (hunk 2 above)
Comment (superm1) [reply plugins/amdgpu/fu-amdgpu-device.c:11]: Even though the kernel driver is 'amdgpu'?

File: plugins/amdgpu/fu-amdgpu-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/amdgpu/fu-amdgpu-device.c:40]: `video-display` IMHO.

File: libfwupdplugin/fu-udev-device.c (hunk 1 above)
Comment (superm1) [reply libfwupdplugin/fu-udev-device.c]: added

File: plugins/amdgpu/fu-amdgpu-device.c (hunk 3 above)
Comment (superm1) [reply plugins/amdgpu/fu-amdgpu-device.c:40]: fixed

File: plugins/amd-pmc/fu-amd-pmc-device.c [hunk 4]
```diff
@@ -59,7 +59,7 @@ fu_amd_pmc_device_init(FuAmdPmcDevice *self)
 	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_INTERNAL);
 	fu_device_add_icon(FU_DEVICE(self), "computer");
 	fu_device_add_parent_guid(FU_DEVICE(self), "cpu");
-	fu_device_set_vendor(FU_DEVICE(self), "AMD");
+	fu_device_set_vendor(FU_DEVICE(self), "Advanced Micro Devices, Inc.");
```
Comment (hughsie) [inline plugins/amd-pmc/fu-amd-pmc-device.c:62]: Can we drop the .Inc?

File: libfwupdplugin/fu-udev-device.h [hunk 5]
```diff
@@ -75,6 +125,10 @@ void
 fu_udev_device_set_bind_id(FuUdevDevice *self, const gchar *bind_id);
 const gchar *
 fu_udev_device_get_driver(FuUdevDevice *self);
+gboolean
+fu_udev_device_is_pci_base_class(FuUdevDevice *self, FuPciBaseClass cls);
+guint32
```
Comment (hughsie) [inline libfwupdplugin/fu-udev-device.h:130]: guint8, or FuPciBaseClass perhaps?

File: plugins/amdgpu/fu-amdgpu-device.c (hunk 2 above)
Comment (hughsie) [reply plugins/amdgpu/fu-amdgpu-device.c:11]: I think so; I think we're settling on `vendor-thing` as plugin names -- for instance the Intel-GSC plugin is `igsc` in the kernel.

File: libfwupdplugin/fu-udev-device.h (hunk 5 above)
Comment (superm1) [reply libfwupdplugin/fu-udev-device.h:130]: It's actually intentional that it returns the whole class, so if we have a need later we can add subclasses

File: plugins/amd-pmc/fu-amd-pmc-device.c (hunk 4 above)
Comment (hughsie) [reply plugins/amd-pmc/fu-amd-pmc-device.c:62]: Okay.

File: libfwupdplugin/fu-udev-device.h (hunk 5 above)
Comment (superm1) [reply libfwupdplugin/fu-udev-device.h:130]: sure

File: plugins/amdgpu/fu-amdgpu-device.c (hunk 2 above)
Comment (superm1) [reply plugins/amdgpu/fu-amdgpu-device.c:11]: OK.

File: libfwupdplugin/fu-udev-device.c [hunk 6]
```diff
@@ -978,6 +1013,43 @@ fu_udev_device_get_number(FuUdevDevice *self)
...
+ * Determines whether the device matches a given pci base class type
+ *
+ * Since: 1.8.11
+ **/
+gboolean
+fu_udev_device_is_pci_base_class(FuUdevDevice *self, FuPciBaseClass cls)
+{
+	FuUdevDevicePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_UDEV_DEVICE(self), FALSE);
+	return !!(priv->class & (cls << 16));
```
Comment (hughsie) [inline libfwupdplugin/fu-udev-device.c:1030]: I think `> 0` is a bit clearer here.

File: libfwupdplugin/fu-udev-device.h [hunk 7]
```diff
@@ -55,6 +55,56 @@ typedef enum {
...
+	FU_PCI_BASE_CLASS_SERIAL_BUS,
+	FU_PCI_BASE_CLASS_WIRELESS,
+	FU_PCI_BASE_CLASS_INTELLIGENT_IO,
+	FU_PCI_BASE_CLASS_SATELLITE,
+	FU_PCI_BASE_CLASS_ENCRYPTION,
+	FU_PCI_BASE_CLASS_SIGNAL_PROCESSING,
+	FU_PCI_BASE_CLASS_ACCELERATOR,
+	FU_PCI_BASE_CLASS_NON_ESSENTIAL,
+	FU_PCI_BASE_CLASS_UNDEFINED = 0xff
+} FuPciBaseClass;
```
Comment (hughsie) [inline libfwupdplugin/fu-udev-device.h:106]: Not sure about the prefix; how about `FuUdevDevicePciBase` -- using _Class_ is a recipe for disaster I think

File: libfwupdplugin/fu-udev-device.c [hunk 8]
```diff
@@ -978,6 +1013,43 @@ fu_udev_device_get_number(FuUdevDevice *self)
...
+ * Gets the PCI class for a device.
+ *
+ * The class consists of a base class and subclass.
+ *
+ * Returns: a PCI class
+ *
+ * Since: 1.8.11
+ **/
+guint32
+fu_udev_device_get_class(FuUdevDevice *self)
```
Comment (hughsie) [inline libfwupdplugin/fu-udev-device.c:1046]: can we avoid the _class suffix?

File: plugins/amd-gpu/amd-gpu.quirk [hunk 9]
```diff
@@ -0,0 +1,2 @@
+[PCI\DRIVER_amdgpu]
+Plugin = amdgpu
```
[the remark this answers is not in the document]
Comment (superm1) [reply plugins/amd-gpu/amd-gpu.quirk:2]: No like I said the kernel driver is 'amdgpu'

File: libfwupdplugin/fu-udev-device.h (hunk 7 above)
Comment (superm1) [reply libfwupdplugin/fu-udev-device.h:106]: How about CLS? I'm worried it really is a class and we want to indicate that

File: libfwupdplugin/fu-udev-device.c (hunk 8 above)
Comment (superm1) [reply libfwupdplugin/fu-udev-device.c:1046]: _cls maybe?

**Review by hughsie — APPROVED** [review]

**Pull request merged**